File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* root, long left, long right){

            if(root==NULL) return true;

            if(root->val > left && root->val < right)
                return helper(root->left, left, root->val) && helper(root->right, root->val, right);  

            return false;
        }

        bool isValidBST(TreeNode* root) {
            if(root==NULL) return true;

            return helper(root, LONG_MIN, LONG_MAX);
        }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root,Long.MIN_VALUE,Long.MAX_VALUE);
    }
   public boolean helper(TreeNode root,long minval,long maxval)
    {
        if(root==null)
        {
            return true;
        }
        if(root.val>=maxval || root.val<=minval)
        {
            return false;
        }
        return helper(root.left,minval,root.val) && helper(root.right,root.val,maxval);
        
    }
    
}